walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,132 @@
+import java.util.List;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+
+  // Returns available times for meeting requesters based on the events and its attendees' schedules.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> optionalAttendees = new HashSet(request.getOptionalAttendees());
+    Set<String> allAttendees = getAllAttendees(events);
+    List<TimeRange> availableTimeRanges = new ArrayList();
+
+    List<Event> eventsListOfMandatory = createSortedEventsListOfMandatory(events, optionalAttendees);
+    List<Event> eventsListOfMandatoryAndOptional = createSortedEventsListOfMandatoryAndOptional(events);
+
+    if (isRequestDurationTooLong(request)) {
+      return availableTimeRanges;
+    }
+
+    if (hasNoOverlappingAttendees(request.getAttendees(), allAttendees)) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return availableTimeRanges;
+    }
+
+    List<TimeRange> timeRangesList = findTimeRangesOfEvents(eventsListOfMandatoryAndOptional);
+    removeNestedAndOverlappedEvents(timeRangesList);
+    addAvailableTimeRanges(availableTimeRanges, timeRangesList, request);
+
+    if (availableTimeRanges.size() == 0) {
+      List<TimeRange> timeRangesListOnlyMandatory = findTimeRangesOfEvents(eventsListOfMandatory);
+      removeNestedAndOverlappedEvents(timeRangesListOnlyMandatory);
+      addAvailableTimeRanges(availableTimeRanges, timeRangesListOnlyMandatory, request);
+    }
+    return availableTimeRanges;
+  }
+
+  // Returns a set of all of the attendees of events.
+  private Set<String> getAllAttendees(Collection<Event> events) {
+    Set<String> allAttendees = new HashSet();
+    for (Event event: events) {
+      allAttendees.addAll(event.getAttendees());
+    }
+    return allAttendees;
+  }
+
+  // Checks if the requesters of a meeting and the attendees have overlapping members.
+  private boolean hasNoOverlappingAttendees(Collection<String> requesters, Set<String> attendees) {
+    return Collections.disjoint(requesters, attendees);
+  }
+
+  // Sorts the TimeRanges of events of only mandatory attendees by their start time.
+  private List<Event> createSortedEventsListOfMandatory(Collection<Event> events, Set<String> optionalAttendees) {
+    List<Event> eventsList = new ArrayList(events);
+    eventsList.removeIf(event -> event.getAttendees().equals(optionalAttendees));
+    return eventsList;
+  }
+
+  // Sorts the TimeRanges of events of both mandatory and optional attendees by their start time.
+  private List<Event> createSortedEventsListOfMandatoryAndOptional(Collection<Event> events) {
+    List<Event> eventsList = new ArrayList(events);
+    Collections.sort(eventsList, Comparator.comparing(Event::getWhen, TimeRange.ORDER_BY_START));
```

Comment:
I think you want to remove if (!event.attendees intersects required attendees) && (!event.attendees intersects optional attendees)

?

Reply:
Hm, I did not remove anything because this function is creating a sorted list of events based on the TimeRange's start time.